site/bps/BP-26-move-distributedlog-core-library.md

```diff
@@ -0,0 +1,60 @@
+---
+title: "BP-26: Move distributedlog library as part of bookkeeper"
+issue: https://github.com/apache/bookkeeper/1024
+state: 'Under Discussion'
+release: "N/A"
+---
+
+### Motivation
+
+DistributedLog is an extension of Apache BookKeeper, which offers *reopenable* log streams as its storage primitives.
+It is tightly built over bookkeeper ledgers, and provides an easier-to-use abstraction and api to use. Applications
+can use *named* log streams rather than *numbered* ledgers to store their data. For example, users can use log streams
+as files to storge objects, checkpoints and other more general filesystem related use cases.
+
+Moving the distributedlog core library as part of bookkeeper would have following benefits:
+
+- It provides more generic "reopenable" log abstraction. It lowers the barrier for people to use bookkeeper to store
+  data, and bring in more use cases into bookkeeper ecosystem.
+- Using ledgers to build continous log stream has been a pattern that been reimplemented multiple times at multiple places,
+  from older projects like HDFS namenode log manager, Hedwig to the newer projects like DistributedLog and Pulsar.
+- Most of the distributedlog usage is using the distributedlog library which only depends Apache BookKeeper and there is no
+  additional components introduced. To simplify those usages, it is better to release distributedlog library along with
+  bookkeeper. It provides a better integration and release procedure.
+
+This proposal proposes "moving the distributedlog library code base as part of bookkeeper and continuing the library
+development in bookkeeper".
+
+### Public Interfaces
+
+This is a new library moved in bookkeeper. It will *NOT* touch any existing bookkeeper modules and ledger api.
+
+### Proposed Changes
+
+This proposal will *ONLY* move following library-only modules from distributedlog repo:
+
+- distributedlog-core: the log stream library that build over bookkeeper ledgr api. It doesn't introduce any service components. Library only.
+- distributedlog-io/dlfs: A hdfs filesystem api wrapper over the log stream api, to provide filesystem-like usage over bookkeeper.
+
+This proposal will *NOT* move other service components like "distributedlog-proxy".
```

Comment:
I'm unclear on what you mean when you say you'll *ONLY* move the above modules, and *NOT* move the following modules. I assume the whole code base is coming in, no?

Reply:
@ivankelly no this proposal is only proposing moving the library not the proxy service for a few reasons. 1) what bookkeeper needs is a *distributedlog*-ish library to simplify the usage of bookkeeper. *ledger* api is not usable for most of the users. 2) moving the library is to mimic the impact of other bookkeeper users who don't need dlog. 3) the dlog proxy was written in twitter-finagle, which depends on a specialized libthrift which is only available in twitter maven. so whether to move dlog proxy or improve it with grpc is a broader discussion than just moving the dlog library. I would defer that in a later discussion.

Hope this makes things clear.

Reply:
So there will still be a distributedlog repo, which will continue to have the proxy service?

AFAIK, we do have other things which are depending on finagle (stats provider i think), and thereby depend on the custom libthrift (which is in public repos now). The finagle team were talking about moving to latest apache libthrift just before christmas.

Reply:
> there will still be a distributedlog repo, which will continue to have the proxy service?

yes. 

> we do have other things which are depending on finagle (stats provider i think), 

stats provider doesn't depend on finagle-thrift.

Reply:
It's twitter-http server, via zipkin-core

Reply:
@ivankelly 

1) twitter server wraps all the utilities such as logging, metrics, tracing. they are dependencies but not necessarily enabled at runtime. and twitter-http-server is one implementation of http-server, it is only used when you load twitter-http server in the classpath.

2) it is a bit different story in distributedlog-proxy uses finagle-thrift as the wire protocol. Unfortunately finagle-thrift depends on a specialized protocol called thrift-mux for multiplexing requests, which only available at twitter's folk, not at the official thrift library. so it is a bit hard to extend to other language bindings. there were multiple issues reported in DL when people tried to create different language bindings, and it is hard to address at current finagle-thrift protocol. there was also discussion on supporting a gRPC proxy, which make multiple language bindings much easier. so it would be good to think about proxy part for different factors. that's why I didn't include proxy in this BP to keep things easier.

Hope this make you clear.

Reply:
fyi - finagle thrift mux - https://twitter.github.io/finagle/guide/Protocols.html

Reply:
Ah, I thought the problem was pulling the custom libthrift in just as a dep.